app/tests/integration/conftest.py

```diff
@@ -0,0 +1,294 @@
+def openstack_connection_fixture(cloud_name: str) -> Connection:
+    """The openstack connection instance."""
+    return openstack.connect(cloud_name)
+
+
+@pytest.fixture(scope="module", name="callback_result_path")
+def callback_result_path_fixture() -> Path:
+    """The file created when the callback script is run."""
+    return Path("callback_complete")
+
+
+@pytest.fixture(scope="module", name="callback_script")
+def callback_script_fixture(callback_result_path: Path) -> Path:
+    """The callback script to use with the image builder."""
+    callback_script = Path("callback")
+    callback_script.write_text(
+        f"""#!/bin/bash
+IMAGE_ID=$1
+echo $IMAGE_ID | tee {callback_result_path}
+""",
+        encoding="utf-8",
+    )
+    callback_script.chmod(0o775)
+    return callback_script
+
+
+@pytest.fixture(scope="module", name="dockerhub_mirror")
+def dockerhub_mirror_fixture(pytestconfig: pytest.Config) -> urllib.parse.ParseResult | None:
+    """Dockerhub mirror URL."""
+    dockerhub_mirror_url: str | None = pytestconfig.getoption("--dockerhub-mirror", default=None)
+    if not dockerhub_mirror_url:
+        return None
+    parse_result = urllib.parse.urlparse(dockerhub_mirror_url)
+    assert (
+        parse_result.netloc and parse_result.port and parse_result.geturl()
+    ), "Invalid dockerhub-mirror URL"
+    return parse_result
+
+
+@pytest.fixture(scope="module", name="openstack_image_name")
+def openstack_image_name_fixture(test_id: str) -> str:
+    """The image name to upload to openstack."""
+    return f"image-builder-test-image-{test_id}"
+
+
+@pytest.fixture(scope="module", name="ssh_key")
+def ssh_key_fixture(
+    openstack_connection: Connection, test_id: str
+) -> typing.Generator[types.SSHKey, None, None]:
+    """The openstack ssh key fixture."""
+    keypair: Keypair = openstack_connection.create_keypair(f"test-image-builder-keys-{test_id}")
+    ssh_key_path = Path("tmp_key")
+    ssh_key_path.touch(exist_ok=True)
+    ssh_key_path.write_text(keypair.private_key, encoding="utf-8")
+
+    yield types.SSHKey(keypair=keypair, private_key=ssh_key_path)
+
+    openstack_connection.delete_keypair(name=keypair.name)
+
+
+@pytest.fixture(scope="module", name="openstack_metadata")
+def openstack_metadata_fixture(
+    openstack_connection: Connection,
+    ssh_key: types.SSHKey,
+    network_name: str,
+    flavor_name: str,
+    cloud_name: str,
+) -> types.OpenstackMeta:
+    """A wrapper around openstack related info."""
+    return types.OpenstackMeta(
+        connection=openstack_connection,
+        cloud_name=cloud_name,
+        ssh_key=ssh_key,
+        network=network_name,
+        flavor=flavor_name,
+    )
+
+
+@pytest.fixture(scope="module", name="openstack_security_group")
+def openstack_security_group_fixture(openstack_connection: Connection, test_id: str):
+    """An ssh-connectable security group."""
+    security_group_name = f"github-runner-image-builder-test-security-group-{test_id}"
```

Comment:
added the test_id here as I've also seen race conditions with parallel test execution (the group gets deleted at the end)